### PR TITLE
Feature/159 subduction rupt builder

### LINKED ISSUE #159: Create Subduction rupture builder (java) and run_ script (python)
So that we have consistent tooling for all rupture build types and can 
run Subduction rupture builds for the project

- [x] update Hikurange Kermadec Fault model with complete slip rates
- [x] new SubductionRupture builder (java)
- [x] add to PythonGateway
- [x] build run script

### PR DESCRIPTION
**Pull request by chrisbc**
closes #159

### CODE REVIEW

File: src/python/automation/scaling/toshi_api.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+import copy
+
+class ToshiApi(ToshiClientBase):
+
+    def __init__(self, url, s3_url, auth_token, with_schema_validation=True, headers=None ):
+        super(ToshiApi, self).__init__(url, auth_token, with_schema_validation, headers)
+        self._s3_url = s3_url
+
+    def OLD_get_general_task_subtask_files(self, id):
+        raise("Don't use this, its too slow, use ")
```
Comment (voj) [inline src/python/automation/scaling/toshi_api.py:12]: should this be deleted then?

Comment (voj) [reply src/python/automation/scaling/toshi_api.py:12]: and what should we use instead?

**Review by voj — COMMENTED** [review]
Looks good. Super long! I've got a few comments

**Comment by voj** [comment]
Just realised, the python code should be in `src/main/python`, not `src/python`

File: src/integration/java/nz/cri/gns/NZSHM22/opensha/hazard/NZSHM22_HazardCalculatorBuilder_IntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package nz.cri.gns.NZSHM22.opensha.hazard;
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply src/integration/java/nz/cri/gns/NZSHM22/opensha/hazard/NZSHM22_HazardCalculatorBuilder_IntegrationTest.java:1]: I guess we are - let's go with it!

File: src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_AbstractInversionRunner.java [hunk 3]
```diff
@@ -54,55 +46,45 @@
 	private CompletionCriteria completionCriteria;
 	private ThreadedSimulatedAnnealing tsa;
 	private double[] initialState;
-	private InversionFaultSystemSolution solution;
-
+	private NZSHM22_InversionFaultSystemSolution solution;
+	
+	private Map<String, Double> finalEnergies = new HashMap<String, Double>();
+	private InversionInputGenerator inversionInputGenerator;
+	protected NZSHM22_SubductionInversionConfiguration inversionConfiguration;
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_AbstractInversionRunner.java:53]: agreed - I've removed this

File: src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_AbstractInversionRunner.java [hunk 4]
```diff
@@ -292,64 +205,21 @@ public NZSHM22_InversionRunner setSlipRateConstraint(
 		this.slipRateConstraintWt_unnormalized = unnormalizedWt;
 		return this;
 	}
-	
-	
+
 	/**
-	 * New NZSHM22 Slip rate uncertainty constraint
+	 * UCERF3 Slip rate uncertainty constraint
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_AbstractInversionRunner.java:210]: Yep, sort of but it's badly worded. I've improved it now I think.

File: src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_CrustalInversionRunner.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	/**
+	 * New NZSHM22 Slip rate uncertainty constraint
+     * @param weightingType
+     * @param uncertaintyWeight
+     * @param scalingFactor
+     * @return
+     */
+    public NZSHM22_CrustalInversionRunner setSlipRateUncertaintyConstraint(String weightingType, 
+			int uncertaintyWeight, int scalingFactor) {		
+    	setSlipRateUncertaintyConstraint(SlipRateConstraintWeightingType.valueOf(weightingType),
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_CrustalInversionRunner.java:89]: oh yes, done!

File: src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_SubductionInversionTargetMFDs.java [hunk 6]
```diff
@@ -73,15 +73,12 @@
 	/*
 	 * MFD constraint default settings
 	 */
-	protected double totalRateM5 = 3d; //5d; 40;
-	protected double bValue = 1d;
+	protected double totalRateM5 = 0.7d; //5d; 40;
+	protected double bValue = 1.1d;
 	protected double mfdTransitionMag = 7.85; // TODO: how to validate this number for NZ? (ref Morgan Page in
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_SubductionInversionTargetMFDs.java:78]: Yes it's been bouncing around a bit. But this is the right place now I believe. These are the 'default' settings for subduction MFDs, and there's a method that allows user to over-ride (which I've just started using). 

For the crustal Target MFDs there'll be different values again.

File: src/python/automation/build_manual_index.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
+#build_rupture_set_index
+
+"""
+Simple script to creawte valid URLs to the rupture sets built
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply src/python/automation/build_manual_index.py:4]: I'll pick this up in another PR.

File: src/python/automation/scaling/toshi_api.py (hunk 1 above)
Comment (chrisbc) [reply src/python/automation/scaling/toshi_api.py:12]: Another PR for this too.

File: src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_AbstractInversionRunner.java [hunk 8]
```diff
@@ -292,64 +205,21 @@ public NZSHM22_InversionRunner setSlipRateConstraint(
...
 	 * 
-	 * @param uncertaintyWeight
-	 * @param scalingFactor
+	 * @param weightingType  a string value from  fromUCERF3InversionConfiguration.SlipRateConstraintWeightingType
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply src/main/java/nz/cri/gns/NZSHM22/opensha/inversion/NZSHM22_AbstractInversionRunner.java]: sorted

**Pull request merged**